content/blog/2022-11-24-announcing-axum-0-6-0.md

```diff
@@ -0,0 +1,346 @@
+
+```rust
+let app = Router::new()
+    // A route on the outermost router that requires `OuterState` as the
+    // state
+    .route("/", get(|_: State<OuterState>| { ... }))
+    // Nest a router under `/api` that requires an `ApiState`
+    // we have to provide the state when nesting it into another router
+    .nest("/api", api_router().with_state(ApiState {}))
+    // Same goes for routers we merge
+    .merge(some_other_routes().with_state(SomeOtherState {}))
+    // Finally provide the `OuterState` needed by the first route we
+    // added
+    .with_state(OuterState {});
+
+// We don't need to provide the state when constructing the sub routers
+//
+// We only need to do that when putting everything together. That means
+// we don't need to pass the different states around to each function
+// that builds a sub router
+fn api_router() -> Router<ApiState> {
+    Router::new()
+        .route("/users", get(|_: State<ApiState>| { ... }))
+}
+
+fn some_other_state() -> Router<SomeOtherState> {
+    Router::new()
+        .route("/foo", get(|_: State<SomeOtherState>| { ... }))
+}
+
+#[derive(Clone)]
+struct ApiState {};
+
+#[derive(Clone)]
+struct SomeOtherState {};
+
+#[derive(Clone)]
+struct OuterState {};
+```
+
+While `Extension` still works we recommend users migrate to `State`, both
+because it is more type safe but also because it is faster.
+
+## Type safe extractor ordering
+
+Continuing the theme of type safety, `axum` now enforces that only one extractor
+consumes the request body. In 0.5 this would compile just fine but fail at
+runtime:
+
+```rust
+use axum::{
+    Router,
+    Json,
+    routing::post,
+    body::Body,
+    http::Request,
+};
+
+let app = Router::new()
+    .route("/", post(handler).get(other_handler));
+
+async fn handler(
+    // This would consume the request body
+    json_body: Json<serde_json::Value>,
+    // This would also attempt to consume the body but fail
+    // since it is gone
+    request: Request<Body>,
+) {}
+
+async fn other_handler(
+    request: Request<Body>,
+    // This would also fail at runtime, even if the `AppState` extension
+    // was set since `Request<Body>` consumes all extensions
+    state: Extension<AppState>,
+) {}
+```
+
+The solution was to manually ensure that you only use one extractor that
+consumes the request and that it was the last extractor. `axum` 0.6 now enforces
+this at compile time:
+
+```rust
+use axum::{
+    Router,
+    Json,
+    routing::post,
+    body::Body,
+    http::Request,
+};
+
+let app = Router::new()
+    .route("/", post(handler).get(other_handler));
+
+async fn handler(
+    // We cannot extract both `Request` and `Json`, have to pick one
+    json_body: Json<serde_json::Value>,
+) {}
+
+async fn other_handler(
+    state: Extension<AppState>,
+    // `Request` must be extracted last
+    request: Request<Body>,
+) {}
+```
+
+This was done by reworking the [`FromRequest`] trait and adding a new
+[`FromRequestParts`] trait.
```

Comment:
I think a short note that people who previously implemented `FromRequest` in a way that doesn't touch the body should move to `FromRequestParts` would be helpful.